### DISCUSSION TITLE
Failed to create alembic chaser in mayaUSDExport command

### DISCUSSION THREAD

**Discussion by mhamid3d** [Autodesk/maya-usd#1304] [Q&A]
I am exporting a simple geometry from Maya using python. I needed to export custom attributes as primvars so I used chasers, but when I add those args to the command, it results in an error, with no other information. I checked the source code at the line it mentioned, it looks like its failing to load the mayaUsdRegistry plugin, but I have no log or anything to work with, so how can this be debugged?

This is the full error:

```
# Error: Failed to create chaser: alembic -- Runtime Error in _BeginWriting at line 404 of /home/S/jenkins/workspace/ecg-mayausd-full-2022-python3-linux/ecg-maya-usd/maya-usd/lib/mayaUsd/fileio/jobs/writeJob.cpp # 
```

The command I used to export the geometry was:

```
import maya.cmds as mc

mc.mayaUSDExport(
    file="/home/mhamid/Downloads/usd_test/walkman2.usd",
    chaser=['alembic'],
    chaserArgs=['alembic', 'primvarprefix', 'CMT_'],
    exportDisplayColor=True,
    shadingMode=None
)
```

Specs:
 - CentOs 8
 - Maya 2022 (Beta PR122)
 - MayaUSD binaries provided for version 0.7.0

**Comment by ysiewappl** [comment]
Hi @mhamid3d:

I ran into this as well previously when working on the import chaser functionality. Please see: https://github.com/Autodesk/maya-usd/pull/1104#issue-560251518 Basically due to the way the repository is structured at the moment you need to load the actual Pixar plugin first _before_ loading the Maya-USD plugin for the macro that registers the chaser to get run.

**Comment by kxl-adsk** [comment]
@mhamid3d Let me first comment on available options for exporting custom attributes. There are at least three I know off:
* use built-in support via custom JSON attribute 
* use python callback
* use chaser (approach you have taken)

There are pros and cons to each of these options. In case it helps somebody, here is a quick python script to illustrate two first methods:
```
from maya import cmds
from pxr import Sdf, Tf, Usd, UsdGeom, Vt

import mayaUsd.lib as mayaUsdLib
import re
import json

USDFILEPATH='/Users/ligenzk/Dev/usd/DeleteMe.usda'
ATTR = {'name':'foo', 'mayaType':'doubleArray', 'mayaValue':[-1., -2., -3., -4., -5., -6.], 'usdInterpolation':UsdGeom.Tokens.faceVarying, 'usdType':Sdf.ValueTypeNames.DoubleArray}

EXPORTED_ATTRS_MAYA_ATTR_NAME = 'USD_UserExportedAttributesJson'

## ---------- UTILS ----------

# Trivial scene with a cube and a single attribute defined above as ATTR dict
def genScene():
    cmds.file(f=True, new=True)
    p = cmds.polyCube()
    s = cmds.listRelatives(p[0], f=True, s=True)
    cmds.addAttr(s[0],ln=ATTR['name'], dt=ATTR['mayaType'])

    a = '{}.{}'.format(s[0],ATTR['name'])
    vals = ', '.join([str(v) for v in ATTR['mayaValue']])
    cmd = 'cmds.setAttr("{}", ({}, {}), type="{}")'.format(a, len(ATTR['mayaValue']), vals, ATTR['mayaType'])
    eval(cmd)

# We should expose UsdMayaUtil::MDagPathToUsdPath to python, but for this example we will go with something simple
def DagPathToSdfPath(longDagPath):
    path = re.sub(r'[|]',r'/',longDagPath)
    sdfPath = Sdf.Path(re.sub(r'[^a-zA-Z0-9_/]',r'_',path))
    return sdfPath

# Method 1 to export custom attributes leveraging JSON description file
def doExportWithJSON():
    # Generate a JSON string to store on a mesh
    jsonDict = {
        ATTR['name'] : { 'interpolation' : ATTR['usdInterpolation'], 'usdAttrName' : ATTR['name'] }
    }
    
    jsonString = json.dumps(jsonDict)
    
    # Set a new attribute with a description of custom attributes to export
    for mesh in cmds.ls(type='mesh',l=True): 
        cmds.addAttr(mesh,ln=EXPORTED_ATTRS_MAYA_ATTR_NAME, dt='string')
        cmds.setAttr('{}.{}'.format(mesh, EXPORTED_ATTRS_MAYA_ATTR_NAME), jsonString, type='string')
        
    cmds.mayaUSDExport(file=USDFILEPATH)

# Method 2 to export custom attributes leveraging python callback function
def doPostExport(filePath):
    exportLayer = Sdf.Layer.Find(filePath)
    exportStage = Usd.Stage.Open(exportLayer, sessionLayer=None)
    
    # Discover attributes you want to export
    for mesh in cmds.ls(type='mesh',l=True):        
        if not ATTR['name'] in cmds.listAttr(mesh):
            continue
        
        sdfPath = DagPathToSdfPath(mesh)
        prim = exportStage.GetPrimAtPath(sdfPath);
        if not prim: # this shape was probably merged
            prim = exportStage.GetPrimAtPath(sdfPath.GetParentPath());
            
        if not prim: # this mesh wasn't exported?
            continue
            
        doubleArrayVt = mayaUsdLib.WriteUtil.GetVtValue('{}.{}'.format(mesh,ATTR['name']),Sdf.ValueTypeNames.DoubleArray)
            
        # Create the primvar and set the value + interpolation type
        prim_pv = UsdGeom.PrimvarsAPI(prim)
        pv1 = prim_pv.CreatePrimvar(ATTR['name'], ATTR['usdType'])
        pv1.Set(doubleArrayVt)
        pv1.SetInterpolation(ATTR['usdInterpolation'])
        pv1.GetAttr().SetCustom(True)

def doExportWithCallback():
    cmds.mayaUSDExport(file=USDFILEPATH, pythonPostCallback="doPostExport('{}')".format(USDFILEPATH))
       
## ---------- MAIN ----------
genScene()

# Option 1 - export using built in support
#doExportWithJSON()
# Option 2 - export using custom python callback
doExportWithCallback()
```

Now, back to your chaser problem. It would help if you could provide a sample project to try. We have regression tests in the repository that are running successfully.

Cheers,
KxL

@ysiewappl - your issue was related to the fact that you made originally the plugin part of pxr binary. I don't believe there is any remaining issue with the test your added - https://github.com/Autodesk/maya-usd/blob/dev/test/lib/usd/plugin/CMakeLists.txt#L138

**Reply by ysiewappl** [reply]
Hi @kxl-adsk: that was only done for the import chaser I added; the Alembic chaser, which I think @mhamid3d was originally asking about, still remains as part of the `pxrUsd` plugin itself, so this problem will still occur: https://github.com/Autodesk/maya-usd/blob/dev/plugin/pxr/maya/plugin/pxrUsd/CMakeLists.txt I think the test also works around the issue by effectively doing the same thing (loading `pxrUsd` rather than `mayaUsd`) https://github.com/Autodesk/maya-usd/blob/dev/plugin/pxr/maya/plugin/pxrUsd/testenv/testPxrUsdAlembicChaser.py

**Reply by kxl-adsk** [reply]
I thought we were talking about a custom chaser (not alembic). For alembic chaser you are absolutely right, it’s part of Pixar plug-in.

**Comment by kxl-adsk** [comment]
I will convert this issue to a discussion until we figure out a way to reproduce the issue with a sample project.
